### ISSUE TITLE
ConnectionResetError in sftp client

### ISSUE THREAD

**Issue by termim** [ronf/asyncssh#272]
Hi Ron,
I have a sftp client downloading several files from Linux server. When I kill the parent sshd process for the corresponding sftp-server processes I get the following exception sometimes:

```python
Traceback (most recent call last):
  ...
  File "C:\PVE\lib\site-packages\asyncssh\sftp.py", line 2091, in get
    await self._begin_copy(self, LocalFile, remotepaths, localpath,
  File "C:\PVE\lib\site-packages\asyncssh\sftp.py", line 1988, in _begin_copy
    await self._copy(srcfs, dstfs, srcfile, dstfile, preserve,
  File "C:\PVE\lib\site-packages\asyncssh\sftp.py", line 1938, in _copy
    await _SFTPFileCopier(block_size, max_requests, 0,
  File "C:\PVE\lib\site-packages\asyncssh\sftp.py", line 436, in run
    if exc and exc.code != FX_EOF:
AttributeError: 'ConnectionResetError' object has no attribute 'code'
```
It seems _SFTPParallelIO.run() doesn't expect *ConnectionResetError* here.
The asyncssh version is 2.2.0.

**Comment by ronf** [comment]
Thanks for the report! That function tests exc.code without first checking if the exception is of type SFTPError. So, if other lower-level failures occur, it could trigger the error you saw.

The following should be a quick fix:

```patch
diff --git a/asyncssh/sftp.py b/asyncssh/sftp.py
index 36de49c..9fc5d92 100644
--- a/asyncssh/sftp.py
+++ b/asyncssh/sftp.py
@@ -433,7 +433,8 @@ class _SFTPParallelIO:
                 for task in done:
                     exc = task.exception()
 
-                    if exc and exc.code != FX_EOF:
+                    if exc and not (isinstance(exc, SFTPError) and
+                                    exc.code == FX_EOF):
                         exceptions.append(exc)
 
                 if exceptions:
```

However, this reminds me that I had been meaning to create subclasses of SFTPError in a manner similar to what I did some time ago for subclasses of DisconnectError, to make it easier to actually provide more specialized exception handling depending on the specific exception code. This seems like a good excuse to do that, as it'll make the above fix a bit cleaner. I'll get that version of the change checked into the "develop" branch soon.

**Comment by ronf** [comment]
Ok - SFTPError subclass support is now available in the "develop" branch, and I have checked in an equivalent to the above fix on top of that in commit f3d7c0f. Thanks again!

**Comment by termim** [comment]
Thanks for the fix! It seems this error is gone with the above changes.

While testing my client on windows I sometimes get a permission error when deleting partially downloaded file after cancelling download on the client side or killing server processes.  It seems that in some cases the output file is not closed before exception is propagated. I've tried to call self.cleanup() before raising exception[0] in _SFTPParallelIO.run() but the problem is not eliminated completely. The problem is that it does not happens all the times download cancelled/connection lost, it depends on in what state client is at this point.

I'm not sure it belongs to this issue, let me know if you want me to open another one.

**Comment by ronf** [comment]
There are various network and file-related errors that could have happened in some cases, and depending on whether those exceptions were converted to an SFTPError exception or not, those could have triggered the "object has no attribute 'code'" error you reported. However. all of those should be gone now with this fix.

Regarding the issue where deleting a file could potentially report back that the file is still open, there are multiple parallel tasks doing I/O on the file, and while _SFTPParallelIO does call cancel() on all of the outstanding tasks before actually raising an exception to the calling code, the current code doesn't actually reap all the canceled tasks before doing so. It counts on the _reap_task() method in the SSHConnection to do this.

That said, there is a "finally" block in _SFTPParallelIO.run() that calls cleanup() automatically when any parallel I/O finishes, and _SFTPFileCopier.cleanup() makes sure that it closes the src & dst files in cleanup() even when the exception is raised. So, even though there might be some tasks left to reap, the file objects should be properly closed in the copy case before the exception is propagated to the caller.

Are you seeing this problem on file copies, or on large parallel reads or writes?

**Comment by termim** [comment]
I've done some more testing and it seems that it happens when cancellation and connection loss step on each other. Below is a script I've used for testing. I run it on Windows 10 machine. It opens 10 sftp sessions to download some 15 MiB file from a Linux box and after some timeout cancels the downloads. In case of cancellation or other exception it tries to delete downloaded files. If I manage to kill couple of sftp server process then I get the PermissionError almost all the time. The capacity between the machines is big, so there is a noticeable timeout between killing the sftp server processes and client noticing that.

```python
# -*- coding: utf-8 -*-

import asyncio
from pathlib import Path
import asyncssh


def progress(src, dst, d, t):
    return
    print(f"{src} -> {dst} {d/t:4}")


async def cancel_it(tasks):
    await asyncio.sleep(4)
    for task in tasks:
        print(f"CANCEL: {task}")
        task.cancel()


async def main():

    host = 'xxxxxxxxxxx'
    user, password = 'xxx', 'xxx'
    fin = "some_big_file"

    tasks = []
    N = 10
    conn = await asyncssh.connect(host=host, username=user, password=password,
                                    known_hosts=None, client_keys=())
    for i in range(N):
        sftp = await conn.start_sftp_client()
        fout = Path(f"/tmp/test.bin{i}")
        tasks.append(asyncio.create_task(sftp.get(fin, fout, progress_handler=progress)))

    await asyncio.ensure_future(cancel_it(tasks))

    done, pending = await asyncio.wait(tasks, return_when=asyncio.ALL_COMPLETED)
    try:
        [ f.result() for f in done ]
    except asyncio.CancelledError:
        print("I'm cancelled")
        for i in range(N):
            fout = Path(f"/tmp/test.bin{i}")
            if fout.exists():
                fout.unlink()
    except Exception as ex:
        print("GOT EXCEPTION: {}".format(ex))
        for i in range(N):
            fout = Path(f"/tmp/test.bin{i}")
            if fout.exists():
                fout.unlink()


if __name__ == '__main__':

    loop = asyncio.get_event_loop()
    #loop.set_debug(True)

    loop.run_until_complete(main())
    print("ALL DONE.")
```

**Comment by ronf** [comment]
Thanks for providing the test code!

Unfortunately, I haven't been able to reproduce the problem here so far. I tried a 16 MB file here and I didn't see the cancellation at all, despite raising the sleep time to 20 seconds to give myself a chance to kill some of the server processes. I have a feeling the transfers all finished before the cancellations began, but the Python process was still waiting for the sleep to finish in cancel_it().

With a larger file, I did manage to get it to report back the 'Connection closed' (or sometimes 'Connection not open'), but I wasn't able to reproduce the permission denied error, even when running on Windows 10. Also, the connection closed error showed up very oddly. It was an unretrieved exception, and it had a HUGE traceback that seemed to involve a bunch of nested calls to _make_request, which I wouldn't expect at all.

**Comment by ronf** [comment]
Looking at the test code more closely here, I can see why it sometimes reports some task exceptions were not retrieved -- it basically stops on the first exception, and never finishes calling f.result() on the rest. I don't fully understand why the stack is so deep in that case, but fiddling with the loop allowed me to avoid all that output and better see what's going on. I still haven't been able to reproduce the Permission Denied, though.

**Comment by ronf** [comment]
Ok - I think I _may_ have a lead here. I was finally able to reproduce the problem, and it appears that when it happens the cleanup function attempts to close the network file object (the source), but never attempts to close the local file object (the destination), leaving that file open when your code later tries to unlink the file.

If an exception occurred the network close operation, I could see this happening as it would exit out of the cleanup() function before completing things. I haven't actually been able to confirm that an exception is being thrown in that case as I can't reproduce it reliably enough to see that yet, but I think this is a possible explanation. If that's the case, using nested "finally" blocks might be one way to address this.

**Comment by ronf** [comment]
I've checked in a potential fix for this in commit b6e68c9 in the "develop" branch. I also did some other cleanup around handling of SFTP exceptions in a previous commit. Could you try this version and see if it fixes the issue you're seeing?

**Comment by ronf** [comment]
Something else I noticed about your test code is that it tries to clean up all of the file objects as soon as it hits the first exception. This could be a problem even after the fix I put in here, as some of those other transfers might still be in the process of being canceled. You probably should wait for the result to come back from _all_ the tasks before you try to clean up the files they might have open, or only clean up the files associated with each task individually after retrieving that task's result.

**Comment by termim** [comment]
I think you are right - ignoring all but the first exception is not the right thing to do here. 
In the code below I've retrieved all the exceptions and now I can not reproduce PermissionError.
I do not understand why this caused delay in cleanup though. I'd expect that after 

```python
await asyncio.wait(tasks, return_when=asyncio.ALL_COMPLETED)
```
everything should be settled down. It seems that even if all the tasks are done the clean up actions are delayed sometimes until all the exceptions are retrieved. It is somewhat confusing IMHO.

I've also managed to get huge tracebacks with _make_request calls even after retrieving all exeptions. It happens almost all the time if I wait till client cancel the download and then quickly delete two server processes. 

It is interesting also that it takes about 1200 tasks to do 10 downloads and around 10s to cancel them.

BTW Ive used the develop branch with the latest changes.

```python
# -*- coding: utf-8 -*-

import asyncio
from pathlib import Path
import logging
import asyncssh

log = logging.error


def progress(src, dst, d, t):
    return
    log(f"{src} -> {dst} {d/t:4}")


async def cancel_it(tasks):
    await asyncio.sleep(15)
    log(f"Tasks Number4:  {len(asyncio.all_tasks())}")
    for task in tasks:
        log(f"CANCEL: {task.get_name()}")
        task.cancel()


async def main():

    host = 'xxxxxxxxxxxx'
    user, password = 'xxx', 'xxx'
    fin = "some_big_file"

    log(f"Tasks Number1:  {len(asyncio.all_tasks())}")
    tasks = []
    N = 10
    conn = await asyncssh.connect(host=host, username=user, password=password,
                                    known_hosts=None, client_keys=())
    log(f"Tasks Number2:  {len(asyncio.all_tasks())}")
    for i in range(N):
        sftp = await conn.start_sftp_client()
        fout = Path(f"/tmp/test.bin{i}")
        tasks.append(asyncio.create_task(sftp.get(fin, fout, progress_handler=progress)))

    log(f"Tasks Number3:  {len(asyncio.all_tasks())}")
    await asyncio.ensure_future(cancel_it(tasks))
    log(f"Tasks Number5:  {len(asyncio.all_tasks())}")

    done, pending = await asyncio.wait(tasks, return_when=asyncio.ALL_COMPLETED)
    log(f"Tasks Number6:  {len(asyncio.all_tasks())}")
    for i, f in enumerate(done):
        try:
            f.result()
        except asyncio.CancelledError:
            log(f"{i}. I'm cancelled {f.get_name()}")
            log(f"Tasks Number7:  {len(asyncio.all_tasks())}")
            for i in range(N):
                fout = Path(f"/tmp/test.bin{i}")
                if fout.exists():
                    fout.unlink()
        except Exception as ex:
            log(f"{i}. GOT EXCEPTION: {f.get_name()} {ex}")
            log(f"Tasks Number8:  {len(asyncio.all_tasks())}")
            for i in range(N):
                fout = Path(f"/tmp/test.bin{i}")
                if fout.exists():
                    fout.unlink()
        else:
            log(f"{i}. done")
    log(f"Tasks Number9:  {len(asyncio.all_tasks())}")


if __name__ == '__main__':

    import sys

    logging.basicConfig(format='%(asctime)s: %(message)s', level=logging.ERROR, stream=sys.stdout)

    loop = asyncio.get_event_loop()
    #loop.set_debug(True)

    loop.run_until_complete(main())
    log("ALL DONE.")
```

**Comment by ronf** [comment]
It's good to hear that you aren't seeing the PermissionDenied error after a combination of my changes and retrieving all the exceptions. You may be right that my changes alone (without changing the test code to retrieve all the exceptions) would have been enough to prevent the Permission Denied error. I was initially thinking that even though you performed the "cancel" operation in one of the tasks, that cancellation might have still been in progress on some of the tasks at the point where the first exception was reported, but I guess that would have only been an issue if you used something like the as_completed() iterator or wait() with FIRST_COMPLETED or FIRST_EXCEPTION. With ALL_COMPLETED, I guess it should block until the cancellations have all happened (or the tasks completed for other reasons) regardless of whether you retrieved the other exceptions, giving the "finally" block which calls cleanup() a chance to run. However, not retrieving the other exceptions means getting all those tracebacks, making it hard to see the other log output. So, it's worthwhile to do that regardless.

As for the large number of tasks, that actually makes sense -- each parallel I/O operation (which defaults to a 16 KB block size) is scheduled as a separate task. So, you'd have nearly 1,000 tasks per file transfer if your file was 15 MB in size, or close to 10,000 tasks total if the transfers fully completed. If the cancellation happened prior to the full file being transferred, the numbers would be lower, but still significant. You can increase the block size, but some SFTP servers have fairly low undocumented limits on the amount a single read() call can return, so it's better to keep the block size pretty low. When I last looked, OpenSSH limited reads to no more than 64 KB, and actually failed if the total SFTP message size exceeded 256 KB. See the note I added in https://asyncssh.readthedocs.io/en/latest/api.html#asyncssh.SFTPClient.open.

As for the deeply nested stack, I haven't figured that one out yet, nor can I figure out why there could still be unretrieved exceptions reported with the new test code (and with a similar change I made here yesterday to the test code). I could see it happening if there was a new exception raised in main() inside one of the except blocks, but in my testing yesterday that wasn't what I saw. In fact, I added extra exception blocks in main() to specifically look for that, and I _still_ saw tracebacks showing unretrieved exceptions.

When you were looking at the time it takes to cancel, did you take into account that the cancel_it() task itself was one of the things being waited for, and that it wouldn't even start for 15 seconds due to the sleep at the top?

**Comment by ronf** [comment]
For what it's worth, I tried out your new test code and so far I haven't been able to trigger unretrieved exceptions there, on either Windows or Mac clients.

**Comment by termim** [comment]
By the time to cancel I mean the time from *Tasks Number5* to *Tasks Number6* log message.

May be the difference is that I do download through Internet?

**Comment by ronf** [comment]
The times were very close together in my tests:

2020-03-28 17:29:05,502: Tasks Number1:  1
2020-03-28 17:29:05,948: Tasks Number2:  1
2020-03-28 17:29:07,085: Tasks Number3:  1153
2020-03-28 17:29:22,131: Tasks Number4:  2
2020-03-28 17:29:22,131: CANCEL: Task-11
2020-03-28 17:29:22,131: CANCEL: Task-142
2020-03-28 17:29:22,131: CANCEL: Task-396
2020-03-28 17:29:22,131: CANCEL: Task-707
2020-03-28 17:29:22,131: CANCEL: Task-886
2020-03-28 17:29:22,131: CANCEL: Task-1350
2020-03-28 17:29:22,131: CANCEL: Task-1737
2020-03-28 17:29:22,131: CANCEL: Task-1929
2020-03-28 17:29:22,131: CANCEL: Task-2248
2020-03-28 17:29:22,131: CANCEL: Task-2787
2020-03-28 17:29:22,131: Tasks Number5:  1
2020-03-28 17:29:22,132: Tasks Number6:  1

It took a couple of seconds to finish the initial SSH handshake, but after that there was the expected 15 second delay between Number3 and Number4 and almost no time at all after that for the cancellation or the awaiting of the other tasks to finish. That was when I killed ALL of the sftp server processes, though. When I left some of those sftp-server processes alive, it took a bit longer between Number5 and Number6:

2020-03-29 20:26:18,796: Tasks Number5:  991
2020-03-29 20:26:19,232: Tasks Number6:  9

I imagine this amount of time would depend on how long the remaining file transfers took to finish. So, if you were downloading over a slow link, that could explain the long delay.

**Comment by termim** [comment]
With the latest develop I also see just ~5sec:
```
2020-03-30 18:20:05,477: Tasks Number1:  1
2020-03-30 18:20:06,264: Tasks Number2:  1
2020-03-30 18:20:16,033: Tasks Number3:  1173
2020-03-30 18:20:31,034: Tasks Number4:  1302
2020-03-30 18:20:31,034: CANCEL: Task-11
2020-03-30 18:20:31,035: CANCEL: Task-142
2020-03-30 18:20:31,035: CANCEL: Task-376
2020-03-30 18:20:31,035: CANCEL: Task-783
2020-03-30 18:20:31,035: CANCEL: Task-1284
2020-03-30 18:20:31,035: CANCEL: Task-1770
2020-03-30 18:20:31,035: CANCEL: Task-2266
2020-03-30 18:20:31,035: CANCEL: Task-2762
2020-03-30 18:20:31,036: CANCEL: Task-3280
2020-03-30 18:20:31,036: CANCEL: Task-3806
2020-03-30 18:20:31,042: Tasks Number5:  1299
2020-03-30 18:20:34,710: Tasks Number6:  11
```
It probably depends on the network condition at the moment as well.

Anyway, I guess the problem is solved and we can close the issue.
Thank you very much for you help and support!

**Comment by ronf** [comment]
Yes, it would - it's all about how long the non-killed transfers would take to finish.

I think you're right that we can consider this resolved. I really appreciate all your help putting together the test case for it. Without that, the fix would have been very difficult to find!

I generally leave issues open until the fix actually makes it into a release, as a reminder to post the version the fix goes into, but I don't think there's anything further needed for now. Thanks again!

**Comment by ronf** [comment]
The fixes discussed here are now available in AsyncSSH 2.2.1.
